### PR TITLE
feat: debug functions

### LINKED ISSUE #2174: add a `uri()` method to `Dataset` in rust
also change the `uri()` method in python to use this new thing

the `uri()` should return the exact same string used to construct the Dataset.

### PR DESCRIPTION
**Pull request by wjones127**
These functions provide debug printing on various important structures. They are mainly useful for debugging issues on the dataset.

Since they are mainly for debugging, they are free function in a submodule instead of easily discoverable methods on respective classes.

Also, this includes a drive-by fix: fixes #2174

As dicussed in https://github.com/lancedb/lance/issues/2186

### CODE REVIEW

File: python/python/lance/fragment.py [hunk 1]
```diff
@@ -428,7 +434,7 @@ def metadata(self) -> FragmentMetadata:
         -------
         FragmentMetadata
         """
-        return FragmentMetadata(self._fragment.metadata().json())
```
Comment (wjones127) [inline python/python/lance/fragment.py:431]: This constructor was causing `FragmentMetadata.schema()` to be empty.

**Review by eddyxu — APPROVED** [review]

File: python/src/debug.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+/// Print the Lance schema of a dataset.
+///
+/// This can be used to view the field ids and types in the schema.
+#[pyfunction]
+pub fn print_schema(dataset: &PyAny) -> PyResult<()> {
+    let py = dataset.py();
+    let dataset = dataset.getattr("_ds")?.extract::<Py<Dataset>>()?;
+    let dataset_ref = &dataset.as_ref(py).borrow().ds;
+    let schema = dataset_ref.schema();
+    println!("{:#?}", schema);
```
Comment (westonpace) [inline python/src/debug.rs:21]: Why did you choose for these functions to actually `println` instead of returning a `String` to python?

**Review by westonpace — APPROVED** [review]
This is awesome, I think my only concern is that we have both `print_xyz` methods and `list_xyz` methods.  I think it would be nice if the `print_...` methods returned a string instead of printing to console.

Or, even better, but maybe not so clearly "debug and internal only" would be to return a python data class with some kind of builtin `__repr__` that printed nicely.  This would allow for both programmatic access as well as quick printing.  But maybe we don't want to do this since we don't want people relying on these internal field names.

File: python/src/debug.rs (hunk 2 above)
Comment (wjones127) [reply python/src/debug.rs:21]: It felt natural at the time. I'm fine with making them return a string, but then maybe `print_` is the wrong prefix. What should I use instead?

Comment (wjones127) [reply python/src/debug.rs:21]: `format_fragment` is okay. I'll make it return a string and do that then.

**Pull request merged**